rpcs3/Emu/Cell/PPUTranslator.cpp

```diff
@@ -3833,7 +3833,21 @@ void PPUTranslator::FMADDS(ppu_opcode_t op)
 	const auto a = GetFpr(op.fra);
 	const auto b = GetFpr(op.frb);
 	const auto c = GetFpr(op.frc);
-	const auto result = m_ir->CreateFPTrunc(m_ir->CreateFAdd(m_ir->CreateFMul(a, c), b), GetType<f32>());
+
+	llvm::Value* result;
+	if (m_use_fma)
+	{
+		result = m_ir->CreateCall(get_intrinsic<f32>(llvm::Intrinsic::fma), {
+			m_ir->CreateFPTrunc(a, GetType<f32>()),
+			m_ir->CreateFPTrunc(c, GetType<f32>()),
+			m_ir->CreateFPTrunc(b, GetType<f32>())
```

Comment:
All those single precision operations dont need truncation and usage of accurate FMA, this is because when the operation is performed in double precision it doesnt have any rounding errors. accurate FMA only matters for double precision instructions.

Reply:
On master I get a (very slightly) wrong result for fmadds/fmsubs/fnmadds/fnmsubs on the test I linked in my first post, which the PR also fixes.

PS3: `fmadds  [04],[16],[17]) -> 46ea61cd`
Master: `fmadds  [04],[16],[17]) -> 46ea61cc`
PR: `fmadds  [04],[16],[17]) -> 46ea61cd`

04 is 1.5, 16 is 20000.6, and 17 is FLT_MIN.

That's why I included them for single-precision, but I can remove it if you want (unless I misunderstood).

Reply:
Comment "needs hw testing" for these single-precision instructions, notice that we do not truncate doubles for these "single" precision intructions anywhere else because past hw testing led us to believe realhw actually uses reads the registers suorces and performs the operation in doubles and then truncates the result to single precision which may produce slightly different results then truncating from sources.

Reply:
again sorry if im being dumb, very new to this, but the following also produces `46ea61cc`:
`result = m_ir->CreateFPTrunc(m_ir->CreateCall(get_intrinsic<f64>(llvm::Intrinsic::fma), {a, c, b}), GetType<f32>());`

I am assuming you mean truncate the result rather than all the sources?

Reply:
yes, I mean truncating only the result.

Reply:
ok, that seems to produce a different result than truncating sources.